File: family_os_expo/services/firebaseService.ts
import {
  collection, doc, setDoc, addDoc, updateDoc, deleteDoc,
  getDoc, getDocs, query, where, orderBy, limit,
  onSnapshot, serverTimestamp, Timestamp, type Unsubscribe,
} from 'firebase/firestore';
import {
  createUserWithEmailAndPassword, signInWithEmailAndPassword,
  signOut, updateProfile, sendPasswordResetEmail,
  onAuthStateChanged, type User,
} from 'firebase/auth';
import { auth, db } from './firebase';
import {
  type Child, type FamilyEvent, type FamilyTask,
  type Family, type FamilyMember,
} from '../models';
import { COLLECTIONS } from '../utils/constants';
import { generateInviteCode } from '../utils/helpers';

// ─── Firestore converters ─────────────────────────────────────────────────────

function toDate(v: unknown): Date {
  if (v instanceof Timestamp) return v.toDate();
  if (v instanceof Date) return v;
  return new Date();
}

function childFromDoc(d: any): Child {
  const data = d.data();
  return { ...data, id: d.id, createdAt: toDate(data.createdAt) };
}

function eventFromDoc(d: any): FamilyEvent {
  const data = d.data();
  return {
    ...data, id: d.id,
    startDate: toDate(data.startDate),
    endDate:   data.endDate ? toDate(data.endDate) : undefined,
    createdAt: toDate(data.createdAt),
  };
}

function taskFromDoc(d: any): FamilyTask {
  const data = d.data();
  return {
    ...data, id: d.id,
    dueDate:      data.dueDate      ? toDate(data.dueDate)      : undefined,
    reminderDate: data.reminderDate ? toDate(data.reminderDate) : undefined,
    completedAt:  data.completedAt  ? toDate(data.completedAt)  : undefined,
    createdAt: toDate(data.createdAt),
  };
}

function familyFromDoc(d: any): Family {
  const data = d.data();
  return {
    ...data, id: d.id,
    createdAt: toDate(data.createdAt),
    members: (data.members ?? []).map((m: any) => ({
      ...m, joinedAt: toDate(m.joinedAt),
    })),
  };
}

// ─── Auth ─────────────────────────────────────────────────────────────────────

export const firebaseService = {
  // Auth
  currentUser: () => auth.currentUser,

  onAuthChange: (cb: (u: User | null) => void): Unsubscribe =>
    onAuthStateChanged(auth, cb),

  register: async (email: string, password: string, name: string) => {
    const cred = await createUserWithEmailAndPassword(auth, email, password);
    await updateProfile(cred.user, { displayName: name });
    return cred.user;
  },

  login: (email: string, password: string) =>
    signInWithEmailAndPassword(auth, email, password),

  logout: () => signOut(auth),

  resetPassword: (email: string) => sendPasswordResetEmail(auth, email),

  // ─── Family ────────────────────────────────────────────────────────────────

  createFamily: async (name: string): Promise<Family> => {
    const user = auth.currentUser!;
    const member: FamilyMember = {
      userId: user.uid,
      name: user.displayName ?? 'Parent',
      email: user.email ?? undefined,
      role: 'admin',
      joinedAt: new Date(),
    };
    const ref = doc(collection(db, COLLECTIONS.families));
    const family: Omit<Family, 'id'> = {
      name,
      members: [member],
      inviteCode: generateInviteCode(),
      primaryLanguage: 'he',
      createdAt: new Date(),
    };
    await setDoc(ref, {
      ...family,
      createdAt: serverTimestamp(),
      members: [{ ...member, joinedAt: serverTimestamp() }],
    });
    // Store familyId on user doc
    await setDoc(doc(db, COLLECTIONS.users, user.uid), { familyId: ref.id }, { merge: true });
    return { ...family, id: ref.id };
  },

  getUserFamilyId: async (uid: string): Promise<string | null> => {
    const snap = await getDoc(doc(db, COLLECTIONS.users, uid));
    return snap.exists() ? (snap.data().familyId as string) : null;
  },

  getFamilyByCode: async (code: string): Promise<Family | null> => {
    const q = query(
      collection(db, COLLECTIONS.families),
      where('inviteCode', '==', code.toUpperCase()),
      limit(1),
    );
    const snap = await getDocs(q);
    if (snap.empty) return null;
    return familyFromDoc(snap.docs[0]);
  },

  joinFamily: async (familyId: string) => {
    const user = auth.currentUser!;
    const member: FamilyMember = {
      userId: user.uid,
      name: user.displayName ?? 'Parent',
      email: user.email ?? undefined,
      role: 'parent',
      joinedAt: new Date(),
    };
    const ref = doc(db, COLLECTIONS.families, familyId);
    const snap = await getDoc(ref);
    const existing: FamilyMember[] = snap.data()?.members ?? [];
    await updateDoc(ref, {
      members: [...existing, { ...member, joinedAt: serverTimestamp() }],
    });
    await setDoc(doc(db, COLLECTIONS.users, user.uid), { familyId }, { merge: true });
  },

  watchFamily: (familyId: string, cb: (f: Family | null) => void): Unsubscribe =>
    onSnapshot(doc(db, COLLECTIONS.families, familyId), (snap) =>
      cb(snap.exists() ? familyFromDoc(snap) : null)),

  // ─── Children ──────────────────────────────────────────────────────────────

  watchChildren: (familyId: string, cb: (c: Child[]) => void): Unsubscribe =>
    onSnapshot(
      query(collection(db, COLLECTIONS.families, familyId, COLLECTIONS.children), orderBy('createdAt')),
      (snap) => cb(snap.docs.map(childFromDoc)),
    ),

  addChild: async (familyId: string, child: Omit<Child, 'id'>): Promise<Child> => {
    const ref = doc(collection(db, COLLECTIONS.families, familyId, COLLECTIONS.children));
    await setDoc(ref, { ...child, createdAt: serverTimestamp() });
    return { ...child, id: ref.id };
  },

  updateChild: (familyId: string, child: Child) =>
    updateDoc(doc(db, COLLECTIONS.families, familyId, COLLECTIONS.children, child.id), {
      ...child, createdAt: Timestamp.fromDate(child.createdAt),
    }),

  deleteChild: (familyId: string, childId: string) =>
    deleteDoc(doc(db, COLLECTIONS.families, familyId, COLLECTIONS.children, childId)),

  // ─── Events ────────────────────────────────────────────────────────────────

  watchEvents: (familyId: string, cb: (e: FamilyEvent[]) => void): Unsubscribe =>
    onSnapshot(
      query(
        collection(db, COLLECTIONS.families, familyId, COLLECTIONS.events),
        orderBy('startDate'),
        limit(200),
      ),
      (snap) => cb(snap.docs.map(eventFromDoc)),
    ),

  addEvent: async (familyId: string, event: Omit<FamilyEvent, 'id'>): Promise<FamilyEvent> => {
    const ref = doc(collection(db, COLLECTIONS.families, familyId, COLLECTIONS.events));
    await setDoc(ref, {
      ...event,
      startDate: Timestamp.fromDate(event.startDate),
      endDate: event.endDate ? Timestamp.fromDate(event.endDate) : null,
      createdAt: serverTimestamp(),
    });
    return { ...event, id: ref.id };
  },

  updateEvent: (familyId: string, event: FamilyEvent) =>
    updateDoc(doc(db, COLLECTIONS.families, familyId, COLLECTIONS.events, event.id), {
      ...event,
      startDate: Timestamp.fromDate(event.startDate),
      endDate: event.endDate ? Timestamp.fromDate(event.endDate) : null,
      createdAt: Timestamp.fromDate(event.createdAt),
    }),

  deleteEvent: (familyId: string, eventId: string) =>
    deleteDoc(doc(db, COLLECTIONS.families, familyId, COLLECTIONS.events, eventId)),

  // ─── Tasks ─────────────────────────────────────────────────────────────────

  watchTasks: (familyId: string, cb: (t: FamilyTask[]) => void): Unsubscribe =>
    onSnapshot(
      query(
        collection(db, COLLECTIONS.families, familyId, COLLECTIONS.tasks),
        orderBy('createdAt', 'desc'),
        limit(100),
      ),
      (snap) => cb(snap.docs.map(taskFromDoc)),
    ),

  addTask: async (familyId: string, task: Omit<FamilyTask, 'id'>): Promise<FamilyTask> => {
    const ref = doc(collection(db, COLLECTIONS.families, familyId, COLLECTIONS.tasks));
    await setDoc(ref, {
      ...task,
      dueDate: task.dueDate ? Timestamp.fromDate(task.dueDate) : null,
      createdAt: serverTimestamp(),
    });
    return { ...task, id: ref.id };
  },

  updateTask: (familyId: string, task: FamilyTask) =>
    updateDoc(doc(db, COLLECTIONS.families, familyId, COLLECTIONS.tasks, task.id), {
      ...task,
      dueDate:     task.dueDate     ? Timestamp.fromDate(task.dueDate)     : null,
      completedAt: task.completedAt ? Timestamp.fromDate(task.completedAt) : null,
      createdAt:   Timestamp.fromDate(task.createdAt),
    }),

  completeTask: (familyId: string, taskId: string) =>
    updateDoc(doc(db, COLLECTIONS.families, familyId, COLLECTIONS.tasks, taskId), {
      status: 'done',
      completedAt: serverTimestamp(),
    }),

  snoozeTask: (familyId: string, taskId: string) =>
    updateDoc(doc(db, COLLECTIONS.families, familyId, COLLECTIONS.tasks, taskId), {
      status: 'snoozed',
    }),

  deleteTask: (familyId: string, taskId: string) =>
    deleteDoc(doc(db, COLLECTIONS.families, familyId, COLLECTIONS.tasks, taskId)),
};

File: family_os/lib/services/firebase_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:firebase_messaging/firebase_messaging.dart';
import 'package:uuid/uuid.dart';
import '../models/child.dart';
import '../models/event.dart';
import '../models/family.dart';
import '../models/inbox_message.dart';
import '../models/task.dart';
import '../utils/constants.dart';
import '../utils/helpers.dart';

class FirebaseService {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;
  final FirebaseAuth _auth = FirebaseAuth.instance;
  final _uuid = const Uuid();

  // ─── Auth ─────────────────────────────────────────────────────────────────

  User? get currentUser => _auth.currentUser;
  Stream<User?> get authStateChanges => _auth.authStateChanges();

  Future<UserCredential> signInWithEmail(String email, String password) =>
      _auth.signInWithEmailAndPassword(email: email, password: password);

  Future<UserCredential> registerWithEmail(String email, String password) =>
      _auth.createUserWithEmailAndPassword(email: email, password: password);

  Future<void> signOut() => _auth.signOut();

  Future<void> updateUserProfile(String name, {String? photoUrl}) =>
      _auth.currentUser!.updateDisplayName(name);

  Future<void> sendPasswordReset(String email) =>
      _auth.sendPasswordResetEmail(email: email);

  // ─── Family ───────────────────────────────────────────────────────────────

  Future<Family> createFamily(String name) async {
    final user = currentUser!;
    final code = AppHelpers.generateInviteCode();
    final familyRef = _firestore.collection(AppConstants.familiesCollection).doc();

    final member = FamilyMember(
      userId: user.uid,
      name: user.displayName ?? 'Parent',
      email: user.email,
      photoUrl: user.photoURL,
      role: MemberRole.admin,
      joinedAt: DateTime.now(),
    );

    final family = Family(
      id: familyRef.id,
      name: name,
      members: [member],
      inviteCode: code,
      primaryLanguage: 'he',
      createdAt: DateTime.now(),
    );

    await familyRef.set(family.toFirestore());

    // Save familyId to user doc
    await _firestore
        .collection(AppConstants.usersCollection)
        .doc(user.uid)
        .set({'familyId': familyRef.id, 'name': user.displayName ?? ''}, SetOptions(merge: true));

    return family;
  }

  Future<Family?> getFamilyByInviteCode(String code) async {
    final query = await _firestore
        .collection(AppConstants.familiesCollection)
        .where('inviteCode', isEqualTo: code.toUpperCase())
        .limit(1)
        .get();

    if (query.docs.isEmpty) return null;
    return Family.fromFirestore(query.docs.first);
  }

  Future<void> joinFamily(String familyId) async {
    final user = currentUser!;
    final member = FamilyMember(
      userId: user.uid,
      name: user.displayName ?? 'Parent',
      email: user.email,
      photoUrl: user.photoURL,
      role: MemberRole.parent,
      joinedAt: DateTime.now(),
    );

    await _firestore
        .collection(AppConstants.familiesCollection)
        .doc(familyId)
        .update({
      'members': FieldValue.arrayUnion([member.toMap()])
    });

    await _firestore
        .collection(AppConstants.usersCollection)
        .doc(user.uid)
        .set({'familyId': familyId}, SetOptions(merge: true));
  }

  Stream<Family?> watchFamily(String familyId) => _firestore
      .collection(AppConstants.familiesCollection)
      .doc(familyId)
      .snapshots()
      .map((doc) => doc.exists ? Family.fromFirestore(doc) : null);

  Future<String?> getUserFamilyId(String userId) async {
    final doc = await _firestore
        .collection(AppConstants.usersCollection)
        .doc(userId)
        .get();
    return doc.data()?['familyId'] as String?;
  }

  // ─── Children ─────────────────────────────────────────────────────────────

  Future<Child> addChild(String familyId, Child child) async {
    final ref = _firestore
        .collection(AppConstants.familiesCollection)
        .doc(familyId)
        .collection(AppConstants.childrenCollection)
        .doc();

    final newChild = child.copyWith(id: ref.id, familyId: familyId);
    await ref.set(newChild.toFirestore());
    return newChild;
  }

  Future<void> updateChild(String familyId, Child child) => _firestore
      .collection(AppConstants.familiesCollection)
      .doc(familyId)
      .collection(AppConstants.childrenCollection)
      .doc(child.id)
      .update(child.toFirestore());

  Future<void> deleteChild(String familyId, String childId) => _firestore
      .collection(AppConstants.familiesCollection)
      .doc(familyId)
      .collection(AppConstants.childrenCollection)
      .doc(childId)
      .delete();

  Stream<List<Child>> watchChildren(String familyId) => _firestore
      .collection(AppConstants.familiesCollection)
      .doc(familyId)
      .collection(AppConstants.childrenCollection)
      .orderBy('createdAt')
      .snapshots()
      .map((s) => s.docs.map(Child.fromFirestore).toList());

  // ─── Events ───────────────────────────────────────────────────────────────

  Future<FamilyEvent> addEvent(String familyId, FamilyEvent event) async {
    final ref = _firestore
        .collection(AppConstants.familiesCollection)
        .doc(familyId)
        .collection(AppConstants.eventsCollection)
        .doc();

    final newEvent = event.copyWith(
      id: ref.id,
      familyId: familyId,
      createdBy: currentUser!.uid,
    );
    await ref.set(newEvent.toFirestore());
    return newEvent;
  }

  Future<void> updateEvent(String familyId, FamilyEvent event) => _firestore
      .collection(AppConstants.familiesCollection)
      .doc(familyId)
      .collection(AppConstants.eventsCollection)
      .doc(event.id)
      .update(event.toFirestore());

  Future<void> deleteEvent(String familyId, String eventId) => _firestore
      .collection(AppConstants.familiesCollection)
      .doc(familyId)
      .collection(AppConstants.eventsCollection)
      .doc(eventId)
      .delete();

  Stream<List<FamilyEvent>> watchEvents(
    String familyId, {
    DateTime? from,
    DateTime? to,
    String? childId,
  }) {
    Query query = _firestore
        .collection(AppConstants.familiesCollection)
        .doc(familyId)
        .collection(AppConstants.eventsCollection);

    if (from != null) {
      query = query.where('startDate',
          isGreaterThanOrEqualTo: Timestamp.fromDate(from));
    }
    if (to != null) {
      query = query.where('startDate',
          isLessThanOrEqualTo: Timestamp.fromDate(to));
    }
    if (childId != null) {
      query = query.where('childId', isEqualTo: childId);
    }

    return query
        .orderBy('startDate')
        .snapshots()
        .map((s) => s.docs.map(FamilyEvent.fromFirestore).toList());
  }

  // ─── Tasks ────────────────────────────────────────────────────────────────

  Future<FamilyTask> addTask(String familyId, FamilyTask task) async {
    final ref = _firestore
        .collection(AppConstants.familiesCollection)
        .doc(familyId)
        .collection(AppConstants.tasksCollection)
        .doc();

    final newTask = task.copyWith(
      id: ref.id,
      familyId: familyId,
      createdBy: currentUser!.uid,
    );
    await ref.set(newTask.toFirestore());
    return newTask;
  }

  Future<void> updateTask(String familyId, FamilyTask task) => _firestore
      .collection(AppConstants.familiesCollection)
      .doc(familyId)
      .collection(AppConstants.tasksCollection)
      .doc(task.id)
      .update(task.toFirestore());

  Future<void> completeTask(String familyId, String taskId) => _firestore
      .collection(AppConstants.familiesCollection)
      .doc(familyId)
      .collection(AppConstants.tasksCollection)
      .doc(taskId)
      .update({
    'status': TaskStatus.done.name,
    'completedAt': Timestamp.fromDate(DateTime.now()),
  });

  Future<void> snoozeTask(String familyId, String taskId) => _firestore
      .collection(AppConstants.familiesCollection)
      .doc(familyId)
      .collection(AppConstants.tasksCollection)
      .doc(taskId)
      .update({'status': TaskStatus.snoozed.name});

  Future<void> deleteTask(String familyId, String taskId) => _firestore
      .collection(AppConstants.familiesCollection)
      .doc(familyId)
      .collection(AppConstants.tasksCollection)
      .doc(taskId)
      .delete();

  Stream<List<FamilyTask>> watchTasks(
    String familyId, {
    String? childId,
    TaskStatus? status,
    bool excludeDone = false,
  }) {
    Query query = _firestore
        .collection(AppConstants.familiesCollection)
        .doc(familyId)
        .collection(AppConstants.tasksCollection);

    if (childId != null) {
      query = query.where('childId', isEqualTo: childId);
    }
    if (status != null) {
      query = query.where('status', isEqualTo: status.name);
    }
    if (excludeDone) {
      query = query.where('status', isNotEqualTo: TaskStatus.done.name);
    }

    return query
        .orderBy('createdAt', descending: true)
        .snapshots()
        .map((s) => s.docs.map(FamilyTask.fromFirestore).toList());
  }

  // ─── Inbox ────────────────────────────────────────────────────────────────

  Future<InboxMessage> addInboxMessage(
    String familyId,
    InboxMessage message,
  ) async {
    final ref = _firestore
        .collection(AppConstants.familiesCollection)
        .doc(familyId)
        .collection(AppConstants.inboxCollection)
        .doc();

    final newMsg = InboxMessage(
      id: ref.id,
      familyId: familyId,
      source: message.source,
      content: message.content,
      childId: message.childId,
      isParsed: message.isParsed,
      parsedData: message.parsedData,
      timestamp: message.timestamp,
      createdBy: currentUser!.uid,
    );

    await ref.set(newMsg.toFirestore());
    return newMsg;
  }

  Stream<List<InboxMessage>> watchInbox(String familyId) => _firestore
      .collection(AppConstants.familiesCollection)
      .doc(familyId)
      .collection(AppConstants.inboxCollection)
      .orderBy('timestamp', descending: true)
      .limit(AppConstants.pageSize)
      .snapshots()
      .map((s) => s.docs.map(InboxMessage.fromFirestore).toList());

  // ─── FCM ──────────────────────────────────────────────────────────────────

  Future<String?> getFcmToken() async {
    try {
      return await FirebaseMessaging.instance.getToken();
    } catch (_) {
      return null;
    }
  }

  Future<void> saveFcmToken(String familyId, String token) => _firestore
      .collection(AppConstants.usersCollection)
      .doc(currentUser!.uid)
      .set({'fcmToken': token, 'familyId': familyId}, SetOptions(merge: true));
}
